### PR TITLE
fix: `Filecoin.ChainGetMessage` server error when requesting a signed message

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

As part of https://github.com/ChainSafe/forest/issues/3639

Changes introduced in this pull request:

- fix `Filecoin.ChainGetMessage` server error when requesting a signed message

Payload:
```json
{
    "jsonrpc": "2.0",
    "id": 2,
    "method": "Filecoin.ChainGetMessage",
    "params": [
            {
                "/": "bafy2bzacedpkf2v5kgm4mvvljer6q3s45yfcdzaiyl3xdoivt5m53iv77v76w"
            }
    ]
}
```

Before:
```json
{
    "jsonrpc": "2.0",
    "error": {
        "code": 0,
        "message": "Serialization error for Cbor protocol: RequireLength { name: \"tuple\", expect: 10, value: 2 }"
    },
    "id": 2
}
```

After:
```json
{
    "jsonrpc": "2.0",
    "result": {
        "Version": 0,
        "To": "t410famf46pkqzlieylsxhenre5ajqkutbbrb652rziy",
        "From": "t410fot3vkzzorqg4alowvghvxx4mhofhtazixbm6z2i",
        "Nonce": 26944,
        "Value": "0",
        "GasLimit": 33780427,
        "GasFeeCap": "201100",
        "GasPremium": "200900",
        "Method": 3844450837,
        "Params": "WMTq9dBOAAAAAAAAAAAAAAAAdPdVZy6MDcAt1qmPW9+MO4p5gygAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAC91cGRhdGUgaGVhbHRoYm90XzMxNDE1OV81IHNldCBjb3VudGVyPWNvdW50ZXIrMQAAAAAAAAAAAAAAAAAAAAAA",
        "CID": {
            "/": "bafy2bzaceb2wxksv3na5ziqyay2wemn7tuigp2i4tvudg3jvsuzmdmvcivmzy"
        }
    },
    "id": 2
}
```

Lotus:
```json
{
    "jsonrpc": "2.0",
    "result": {
        "Version": 0,
        "To": "t410famf46pkqzlieylsxhenre5ajqkutbbrb652rziy",
        "From": "t410fot3vkzzorqg4alowvghvxx4mhofhtazixbm6z2i",
        "Nonce": 26944,
        "Value": "0",
        "GasLimit": 33780427,
        "GasFeeCap": "201100",
        "GasPremium": "200900",
        "Method": 3844450837,
        "Params": "WMTq9dBOAAAAAAAAAAAAAAAAdPdVZy6MDcAt1qmPW9+MO4p5gygAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAC91cGRhdGUgaGVhbHRoYm90XzMxNDE1OV81IHNldCBjb3VudGVyPWNvdW50ZXIrMQAAAAAAAAAAAAAAAAAAAAAA",
        "CID": {
            "/": "bafy2bzaceb2wxksv3na5ziqyay2wemn7tuigp2i4tvudg3jvsuzmdmvcivmzy"
        }
    },
    "id": 2
}
```

```console
     Running `target/debug/forest-tool api compare /home/me/fr/snapshots/calibnet/forest_snapshot_calibnet_2023-11-09_height_1073049.forest.car.zst`
| RPC Method                             | Forest        | Lotus         |
|----------------------------------------|---------------|---------------|
| Filecoin.ChainGetBlock                 | Valid         | Valid         |
| Filecoin.ChainGetBlockMessages (73)    | Valid         | Valid         |
| Filecoin.ChainGetGenesis               | Valid         | Valid         |
| Filecoin.ChainGetMessage (4)           | Valid         | Valid         |
| Filecoin.ChainGetMessagesInTipset (20) | Valid         | Valid         |
| Filecoin.ChainGetParentMessages (73)   | MissingMethod | Valid         |
| Filecoin.ChainGetTipSet                | Valid         | Valid         |
| Filecoin.ChainGetTipSetByHeight        | Valid         | Valid         |
| Filecoin.ChainHead                     | Valid         | Valid         |
| Filecoin.ChainReadObj                  | Valid         | Valid         |
| Filecoin.Discover                      | MissingMethod | Valid         |
| Filecoin.MpoolPending                  | Valid         | Valid         |
| Filecoin.NetAddrsListen                | Valid         | Valid         |
| Filecoin.NetInfo                       | Valid         | MissingMethod |
| Filecoin.NetPeers                      | Valid         | Valid         |
| Filecoin.Session                       | MissingMethod | Valid         |
| Filecoin.StartTime                     | Valid         | Valid         |
| Filecoin.StateAccountKey (4)           | MissingMethod | Valid         |
| Filecoin.StateCirculatingSupply (20)   | MissingMethod | Valid         |
| Filecoin.StateDecodeParams (4)         | MissingMethod | Valid         |
| Filecoin.StateGetActor                 | Valid         | Valid         |
| Filecoin.StateGetRandomnessFromBeacon  | MissingMethod | Valid         |
| Filecoin.StateMinerActiveSectors       | MissingMethod | Valid         |
| Filecoin.StateMinerPower (73)          | MissingMethod | Valid         |
| Filecoin.StateNetworkName              | Valid         | Valid         |
| Filecoin.StateReadState                | MissingMethod | Valid         |
| Filecoin.Version                       | Valid         | Valid         |
```

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Review by lemmih — APPROVED** [review]
Can we use `ChainMessage`? A `ChainMessage` is either a `Message` or a `SignedMessage`.

File: src/rpc/chain_api.rs [hunk 1]
```diff
@@ -29,12 +30,17 @@ pub(in crate::rpc) async fn chain_get_message<DB: Blockstore>(
...
 ) -> Result<LotusJson<Message>, JsonRpcError> {
-    let ret: Message = data
-        .state_manager
-        .blockstore()
-        .get_cbor(&msg_cid)?
-        .ok_or("can't find message with that cid")?;
-    Ok(LotusJson(ret))
+    let bytes = data.state_manager.blockstore().get(&msg_cid)?;
+    match bytes {
+        Some(bz) => match fvm_ipld_encoding::from_slice::<Message>(&bz) {
```
Comment (LesnyRumcajs) [inline src/rpc/chain_api.rs:35]: what is `bz`?

Comment (hanabi1224) [reply src/rpc/chain_api.rs:35]: Hmm not sure, let me rename it, it's from `fvm_ipld_encoding` code
```rust
/// Get typed object from block store by Cid.
    fn get_cbor<T>(&self, cid: &Cid) -> anyhow::Result<Option<T>>
    where
        T: de::DeserializeOwned,
    {
        match self.get(cid)? {
            Some(bz) => {
                let res = crate::from_slice(&bz)?;
                Ok(Some(res))
            }
            None => Ok(None),
        }
    }

```

Comment (hanabi1224) [reply src/rpc/chain_api.rs]: Fixed by using `enum ChainMessage` instead

**Comment by hanabi1224** [comment]
> Can we use `ChainMessage`? A `ChainMessage` is either a `Message` or a `SignedMessage`.

@lemmih Yes, that works, fixed

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**